### PR TITLE
Implement max duration of objects in pool

### LINKED ISSUE #5799: Allow specifying the duration an object can stay in a pool
**Jetty version**
9.4.x

**Description**
Currently, the `Pool` class keeps the objects it contains for an indefinite duration. While the number of times a pooled object can be used before it has to be closed can be configured which allows limiting the maximum time an object can stay in its pool, this can seriously shorten their useful lifespan if they are heavily used.

Introducing a new configurable property that specifies exactly how long an object can be pooled before it needs to be cleaned up would guarantee that pooled objects have both a maximum and a minimum lifespan. This would be useful for instance to help failing over HTTP connections in a predictable time to another server after the destination's DNS entry has been modified.

### PR DESCRIPTION
**Pull request by lorban**
Closes #5799

### CODE REVIEW

File: jetty-client/src/main/java/org/eclipse/jetty/client/AbstractConnectionPool.java [hunk 1]
```diff
@@ -105,6 +106,16 @@ protected int getMaxUsageCount()
         return pool.getMaxUsageCount();
     }
 
+    protected void setMaxDuration(long timeInMs)
```
Comment (sbordet) [inline jetty-client/src/main/java/org/eclipse/jetty/client/AbstractConnectionPool.java:109]: Use `java.time.Duration`.

File: jetty-client/src/main/java/org/eclipse/jetty/client/DuplexConnectionPool.java [hunk 2]
```diff
@@ -54,4 +54,17 @@ public void setMaxUsageCount(int maxUsageCount)
...
     }
+
+    @Override
+    public long getMaxDuration()
+    {
+        return super.getMaxDuration();
+    }
+
+    @Override
+    @ManagedAttribute(value = "The maximum duration in milliseconds a connection can be used for before it gets closed")
```
Comment (sbordet) [inline jetty-client/src/main/java/org/eclipse/jetty/client/DuplexConnectionPool.java:65]: JMX annotations go in the getter.
Can you verify if JMC is able to convert to/from a `Duration` object?

File: jetty-util/src/main/java/org/eclipse/jetty/util/Pool.java [hunk 3]
```diff
@@ -205,6 +208,27 @@ public final void setMaxUsageCount(int maxUsageCount)
         copy.forEach(IO::close);
     }
 
+    /**
+     * Get the maximum time span during which the entries of the pool can be acquired.
+     * @param unit the time unit in which to report the time.
+     * @return the duration in the given unit.
+     */
+    public long getMaxDuration(TimeUnit unit)
```
Comment (sbordet) [inline jetty-util/src/main/java/org/eclipse/jetty/util/Pool.java:216]: Use `Duration`, much simpler to do conversions.

File: jetty-client/src/main/java/org/eclipse/jetty/client/MultiplexConnectionPool.java [hunk 4]
```diff
@@ -68,4 +68,17 @@ public void setMaxUsageCount(int maxUsageCount)
...
     }
+
+    @Override
+    public long getMaxDuration()
+    {
+        return super.getMaxDuration();
+    }
+
+    @Override
+    @ManagedAttribute(value = "The maximum duration in milliseconds a connection can be used for before it gets closed")
```
Comment (sbordet) [inline jetty-client/src/main/java/org/eclipse/jetty/client/MultiplexConnectionPool.java:79]: JMX annotation in the getter.

File: jetty-util/src/main/java/org/eclipse/jetty/util/Pool.java [hunk 5]
```diff
@@ -180,6 +182,7 @@ public int getMaxUsageCount()
     /**
      * Change the max usage count of the pool's entries. All existing
      * idle entries over this new max usage are removed and closed.
+     * 0 is illegal and a negative value means there is no limit.
```
Comment (sbordet) [inline jetty-util/src/main/java/org/eclipse/jetty/util/Pool.java:185]: Better: "Value {@code 0} is illegal..."

**Review by sbordet — CHANGES_REQUESTED** [review]
As discussed, let's try to implement expiration at `ConnectionPool` level as suggested by @joakime.

**Review by gregw — CHANGES_REQUESTED** [review]
This looks to be entirely passive, in that an attempt to acquire has to be made before any duration expired entries are detected.  Even then, only those entries attempted will be detected and if all the activity is happening on valid ThreadLocal cached entries, then the rest of the entries might never be checked.

I think we need an active sweeper to find expired entries, otherwise we are just a little more likely to age them than current implementation of a usage count.

But then, having active behaviour doesn't feel like a core Pool thing.  Maybe we need a `TtlPool` extending Pool?

**Comment by lorban** [comment]
@gregw We discussed this together with @joakime and @sbordet and we agreed that the pool class looks like the wrong layer to put this feature at.
This is why there is this 2nd PR that is about implementing the feature at the HTTP connection pool layer, leaving the lowest-layer pool code untouched: https://github.com/eclipse/jetty.project/pull/5801
Let's see what we find best, I'm personally in favor of the 2nd PR.

**Comment by gregw** [comment]
@lorban so this PR is to be closed?  Or is it that the TTL feature is split between layers of abstraction?

**Comment by lorban** [comment]
@gregw I'm unsure yet. My hope is that the TTL feature can be entirely implemented at the HTTP connection pool layer but this has yet to be proven.

If the work goes as planned, this PR can be completely abandoned. Otherwise, we'll have to see what alternatives we have.

**Comment by lorban** [comment]
This is going to be done in the HTTP connection pool instead.

**Pull request closed without merging**